doc/design.md

```diff
@@ -1,158 +1,530 @@
+
+#### 1.1.2 Translating the program into machine code
+
+This step is inevitable because we don't want to limit the algorithm to a specific CPU architecture. In order to generate machine code as fast as possible, we need our instruction set to be as close to native hardware as possible, while still generic enough to support different architectures. There is not enough time for expensive optimizations during code compilation.
+
+#### 1.1.3 Executing the program
+
+The actual program execution should utilize as many CPU components as possible. Some of the features that should be utilized in the program are:
+
+* multi-level caches (L1, L2, L3)
+* μop cache [[6](https://en.wikipedia.org/wiki/CPU_cache#Micro-operation_(%CE%BCop_or_uop)_cache)]
+* arithmetic logic unit (ALU)
+* floating point unit (FPU)
+* memory controller
+* instruction level parallelism [[7](https://en.wikipedia.org/wiki/Instruction-level_parallelism)]
+    * superscalar execution [[8](https://en.wikipedia.org/wiki/Superscalar_processor)]
+    * out-of-order execution [[9](https://en.wikipedia.org/wiki/Out-of-order_execution)]
+    * speculative execution [[10](https://en.wikipedia.org/wiki/Speculative_execution)]
+    * register renaming [[11](https://en.wikipedia.org/wiki/Register_renaming)]
+
+Chapter 2 describes how the RandomX VM takes advantages of these features.
+
+#### 1.1.4 Calculating the final result
+
+Blake2b [[12](https://blake2.net/)] is a cryptographically secure hashing function that was specifically designed to be fast in software, especially on modern 64-bit processors, where it's around three times faster than SHA-3 and can run at a speed of around 3 clock cycles per byte of input. This function is an ideal candidate to be used in a CPU-friendly proof of work.
 
-A CPU-bound algorithm must make use of as many CPU components as possible. Typical CPU contains the following building blocks:
+For processing larger amounts of data in a cryptographically secure way, the Advanced Encryption Standard (AES) [[13](https://en.wikipedia.org/wiki/Advanced_Encryption_Standard)] can provide the fastest processing speed because many modern CPUs support hardware acceleration of these operations. See chapter 3 for more details about the use of AES in RandomX.
 
-* Instruction fetch/decode
-* Branch processing unit
-* Caches
-* Floating point unit
-* Arithmetic logic unit
-* Load-store unit
-* Memory controller
+### 1.2 The "Easy program problem"
 
-Modern CPUs also contain other circuits, such as PCI-E controllers, SATA, USB or Ethernet. Unfortunately, these cannot be used to perform any useful work.
+When a random program is generated, it can be executed only when it's favorable to do so.   This strategy is viable for two main reasons:
```

Comment:
I would say "When a random program is generated, one may choose to execute it only when it's favorable"